Add implementation for data quality checks

marijaselakovic opened the pull request:
## Summary of the changes / Why this is an improvement

- The implementation does data quality checks based on SQLColumnCheck and SQLTableCheck operators
- The implementation requires Airflow 2.3.4 (to be released); with older versions, it doesn't run table checks without throwing an exception
- The implementation sends a Slack message in case some of the data checks failed

## Checklist

 - [ ] [CLA](https://crate.io/community/contribute/cla/) is signed

hammerhead commented on dags/data_quality_checks_dag.py:
```diff
@@ -0,0 +1,153 @@
+    import_data = PostgresOperator.partial(
+        task_id="import_data_to_cratedb",
+        postgres_conn_id="cratedb_connection"
+    ).expand(sql=get_import_statements(S3_BUCKET, "incoming"))
```
Before proceeding with the checks, I think we should do a `REFRESH` on the target table. Due to CrateDB's eventual consistency, it could otherwise be the case that data won't be included in subsequent queries for a short amount of time.

marijaselakovic replied:
I agree with this one, good idea :)

marijaselakovic replied:
Implemented in new commit :)

hammerhead commented on setup/smart_home_data.sql:
```diff
@@ -0,0 +1,37 @@
+   "living_room" DOUBLE PRECISION,
+   "temperature" DOUBLE PRECISION,
+   "humidity" DOUBLE PRECISION
+)
```
```suggestion
);
```
In case someone pastes the two statements into an SQL editor and tries to run them, semicolons will be needed.

hammerhead commented on setup/smart_home_data.sql:
```diff
@@ -0,0 +1,37 @@
+   "living_room" DOUBLE PRECISION,
+   "temperature" DOUBLE PRECISION,
+   "humidity" DOUBLE PRECISION
+)
```
```suggestion
);
```

marijaselakovic commented on dags/data_quality_checks_dag.py:
```diff
@@ -0,0 +1,153 @@
+        task_id='slack_notification',
+        http_conn_id='slack_webhook',
+        message=slack_msg)
+    return failed_alert.execute(context=context)
```
Not sure about this one, I was relying on the Astronomer guide and code for implementing the `on_failure` callback: https://www.astronomer.io/guides/error-notifications-in-airflow/ @denimalpaca is this fine, or should we use another way to implement Slack notification?

denimalpaca replied:
I thought this was bad practice, but after reviewing with the team, it seems doing this is fine.

marijaselakovic commented on dags/data_quality_checks_dag.py:
```diff
@@ -0,0 +1,153 @@
+def get_files_from_s3(bucket, prefix_value):
+    s3_hook = S3Hook()
+    paths = s3_hook.list_keys(bucket_name=bucket, prefix=prefix_value)
+    return paths
```
This is a valid comment and I tried that initially. However, I struggled with imposing the order of tasks modeled with `@task `decorator. In this case, how can I can enforce that `get_files_from_s3` get executed after `move_incoming_files` and `get_import_statements` only after `upload_local_files`:
![Screenshot 2022-08-10 at 13 48 11](https://user-images.githubusercontent.com/7926726/183894090-650a9a8e-1aca-4acd-bb52-cc85bec2d58f.png)

hammerhead replied:
Should be possible like this:
```diff
@@ -52,7 +53,7 @@ def slack_failure_notification(context):
         message=slack_msg)
     return failed_alert.execute(context=context)
 
-
+@task
 def get_files_from_s3(bucket, prefix_value):
     s3_hook = S3Hook()
     paths = s3_hook.list_keys(bucket_name=bucket, prefix=prefix_value)
@@ -144,10 +145,13 @@ with DAG(
                 dest_bucket_key=f"processed-data/{file}",
             )
 
+    files = get_files_from_s3(S3_BUCKET, "incoming")
+    upload_group >> files
+
     delete_landing_files = S3DeleteObjectsOperator.partial(
         task_id='delete_incoming_files',
         aws_conn_id='aws_default',
         bucket=S3_BUCKET
-    ).expand(keys=get_files_from_s3(S3_BUCKET, "incoming"))
+    ).expand(keys=files)
 
 upload_group >> import_data >> data_checks >> move_data >> delete_data >> move_group >> delete_landing_files
```

marijaselakovic replied:
I also added `@task` decorator in the latest commit.

hammerhead commented:
When I started Airflow without setting any of the environment variables used in the DAG, it fails to import:
<img width="661" alt="Screenshot 2022-08-10 at 14 11 11" src="https://user-images.githubusercontent.com/218003/183897927-a9405106-553c-4008-b463-2420d5c8c593.png">

Doesn't make sense of course to run the DAG that way, but maybe there is a more graceful/controlled way of failing/having the DAG don't do anything?

hammerhead commented on dags/data_quality_checks_dag.py:
```diff
@@ -0,0 +1,153 @@
+
+S3_BUCKET = os.environ.get("S3_BUCKET")
+ACCESS_KEY = os.environ.get("ACCESS_KEY")
+SECRET_KEY = os.environ.get("SECRET_KEY")
```
For compatibility with our other DAGs (table export DAG): There we called the environment variables `ACCESS_KEY_ID` and `SECRET_ACCESS_KEY`. That's also the terminology AWS uses.

marijaselakovic replied:
Sure, that one I can add :)

marijaselakovic replied:
This is also fixed in new commit.

marijaselakovic commented:
> When I started Airflow without setting any of the environment variables used in the DAG, it fails to import: <img alt="Screenshot 2022-08-10 at 14 11 11" width="661" src="https://user-images.githubusercontent.com/218003/183897927-a9405106-553c-4008-b463-2420d5c8c593.png">
> 
> Doesn't make sense of course to run the DAG that way, but maybe there is a more graceful/controlled way of failing/having the DAG don't do anything?

This would be nice to have, but currently, I don't have any idea how to control if env variables are not set up properly or if there is another issue with DAG that causes import errors.

hammerhead commented:
`pytest` shows some deprecation warnings: https://github.com/crate/crate-airflow-tutorial/runs/7768590206?check_suite_focus=true#step:5:51

denimalpaca commented on dags/data_quality_checks_dag.py:
```diff
@@ -0,0 +1,173 @@
+from include.data_checks import COL_CHECKS, TABLE_CHECKS
+
+
+S3_BUCKET = os.environ.get("S3_BUCKET")
+ACCESS_KEY_ID = os.environ.get("SECRET_ACCESS_KEY")
+SECRET_ACCESS_KEY = os.environ.get("SECRET_ACCESS_KEY")
+FILE_DIR = os.environ.get("FILE_DIR")
+TEMP_TABLE = os.environ.get("TEMP_TABLE")
+TABLE = os.environ.get("TABLE")
```
Would be nice to include a sample `.env` file with this DAG to make it as simple as possible to get the DAG running.

marijaselakovic replied:
I am not in favor of this as `.env` file contains sensitive data about CrateDB connection and AWS credentials.

denimalpaca replied:
Although looking at this now, seems like `ACCESS_KEY_ID` should get the `access_key` not the secret from the env. Regardless, I'm approving the PR :)

marijaselakovic replied:
This is also fixed in the latest commit :)

marijaselakovic replied:
I see benefits for storing placeholder data in `.env`, but when working with the code locally, there is always a small possibility that someone pushes their own `.env` file (we would need to remove `.env` from `.gitignore`). But, in the comment section of this DAG, I can list which environment variables are required and put placeholders there.

denimalpaca replied:
Ah true, better safe than sorry. Comment should be plenty!

denimalpaca commented on dags/data_quality_checks_dag.py:
```diff
@@ -0,0 +1,173 @@
+
+
+@task
+def get_import_statements(bucket, prefix_value):
+    s3_hook = S3Hook()
+    file_paths = s3_hook.list_keys(bucket_name=bucket, prefix=prefix_value)
```
To reduce redundant calls (and show off the way taskflow works!), you can have this function just take `paths` as a value and use the `get_files_from_s3` function above to retrieve the data.

marijaselakovic replied:
This is done!

denimalpaca commented on dags/data_quality_checks_dag.py:
```diff
@@ -0,0 +1,173 @@
+    start_date=datetime(2021, 1, 1),
+    schedule_interval=None,
+    catchup=False,
+    on_failure_callback=slack_failure_notification
```
Do we want the failure callback to run if *any* task fails, or only particular ones? If it's all of them, then I think this should be under the `default_args` parameter to pass to each task.

marijaselakovic replied:
Done!

denimalpaca commented on dags/data_quality_checks_dag.py:
```diff
@@ -0,0 +1,173 @@
+        trigger_rule='all_done'
+    )
+
+    refresh >>  checks >> move_data >> delete_data
```
nit: extra space

Also a note for readability, I tend to find it easier to group all of the task dependencies together, especially using `chain()` in favor of the bitshift operator.

marijaselakovic replied:
I agree that it makes sense to put all dependencies together (at the end) and did you think here about `set_downstream` and `set_uprstream`?

denimalpaca replied:
Yes, that's the one!

denimalpaca commented on dags/data_quality_checks_dag.py:
```diff
@@ -0,0 +1,173 @@
+                dest_bucket_key=f"processed-data/{file}",
+            )
+
+    s3_files = get_files_from_s3(S3_BUCKET, "incoming")
```
Should have this higher up and let the `get_import_statements()` have it as a dependency as well!

marijaselakovic replied:
Done!

denimalpaca commented on include/data_checks.py:
```diff
@@ -0,0 +1,34 @@
+"""
+Column Level Checks
+"""
+COL_CHECKS = [
+    {"time": {
+        "null_check": {"equal_to": 0},
+        "unique_check": {"equal_to": 0}
+    }},
+    {"use_kw": {
+        "null_check": {"equal_to": 0}
+    }},
+    {"gen_kw": {
+        "null_check": {"equal_to": 0}
+    }},
+    {"temperature": {
+        "min": {"geq_to": -20},
+        "max": {"less_than": 99}
+    }},
+    {"humidity": {
+        "min": {"geq_to": 0},
+        "max": {"less_than": 1}
+    }}
+]
+
+"""
+Table Level Checks
+"""
+TABLE_CHECKS = [
+    {"row_count_check": {"check_statement": "COUNT(*) > 100000"}},
+    {"total_usage_check": {"check_statement": "dishwasher + home_office + "
+                           + "fridge + wine_cellar + kitchen + "
+                           + "garage_door + microwave + barn + "
+                           + " well + living_room  <= house_overall"}}
+]
```
The operators will be much more efficient if all these checks are just one column and one table operators. I think we should prefer demonstrating good techniques in calling databases over another dynamic mapping example.

marijaselakovic replied:
Thanks for the feedback on this. I moved all column mappings and table checks to the operators and they now execute without dynamic task mapping.

denimalpaca reviewed (changes requested):
Looks pretty good overall, left some comments on more minor changes.

marijaselakovic commented:
I have updated the DAG and addressed all comments except the one for `.env` file. I think we should avoid any leakage of sensitive data and that's why I am resisting pushing it.

denimalpaca reviewed (approved):
lgtm with the changes! Only other ask is to run `black` or some other compliant formatter over the files if that hasn't been done already.
